llvmlite/binding/newpassmangers.py

```diff
@@ -0,0 +1,279 @@
+from ctypes import c_bool, c_int
+from llvmlite.binding import ffi
+
+
+def create_new_module_pass_manager():
+    return ModulePassManager()
+
+
+def create_new_function_pass_manager():
+    return FunctionPassManager()
+
+
+def create_pass_builder(tm, pto):
+    return PassBuilder(tm, pto)
+
+
+def create_pipeline_tuning_options(opt_level=2, size_level=0):
+    return PipelineTuningOptions(opt_level=opt_level, size_level=size_level)
+
+
+class ModulePassManager(ffi.ObjectRef):
+
+    def __init__(self, ptr=None):
+        if ptr is None:
+            ptr = ffi.lib.LLVMPY_CreateNewModulePassManager()
+        super().__init__(ptr)
+
+    def run(self, module, pb):
+        ffi.lib.LLVMPY_RunNewModulePassManager(self, pb, module)
+
+    def addVerifier(self):
+        ffi.lib.LLVMPY_AddVerifierPass(self)
+
+    def add_aa_eval_pass(self):
+        ffi.lib.LLVMPY_AddAAEvalPass_module(self)
+
+    def add_simplify_cfg_pass(self):
+        ffi.lib.LLVMPY_AddSimplifyCFGPass_module(self)
+
+    def add_loop_unroll_pass(self):
+        ffi.lib.LLVMPY_AddLoopUnrollPass_module(self)
+
+    def add_loop_rotate_pass(self):
+        ffi.lib.LLVMPY_AddLoopRotatePass_module(self)
+
+    def add_instruction_combine_pass(self):
+        ffi.lib.LLVMPY_AddInstructionCombinePass_module(self)
+
+    def add_jump_threading_pass(self, threshold=-1):
+        ffi.lib.LLVMPY_AddJumpThreadingPass_module(self, threshold)
+
+    def _dispose(self):
+        ffi.lib.LLVMPY_DisposeNewModulePassManger(self)
+
+
+class FunctionPassManager(ffi.ObjectRef):
+
+    def __init__(self, ptr=None):
+        if ptr is None:
+            ptr = ffi.lib.LLVMPY_CreateNewFunctionPassManager()
+        super().__init__(ptr)
+
+    def add_aa_eval_pass(self):
+        ffi.lib.LLVMPY_AddAAEvalPass_function(self)
+
+    def add_simplify_cfg_pass(self):
+        ffi.lib.LLVMPY_AddSimplifyCFGPass_function(self)
+
+    def run(self, fun, pb):
+        ffi.lib.LLVMPY_RunNewFunctionPassManager(self, pb, fun)
+
+    def add_loop_unroll_pass(self):
+        ffi.lib.LLVMPY_AddLoopUnrollPass_function(self)
+
+    def add_loop_rotate_pass(self):
+        ffi.lib.LLVMPY_AddLoopRotatePass_function(self)
+
+    def add_instruction_combine_pass(self):
+        ffi.lib.LLVMPY_AddInstructionCombinePass_function(self)
+
+    def add_jump_threading_pass(self, threshold=-1):
+        ffi.lib.LLVMPY_AddJumpThreadingPass_function(self, threshold)
+
+    def _dispose(self):
+        ffi.lib.LLVMPY_DisposeNewFunctionPassManger(self)
+
+
+class PipelineTuningOptions(ffi.ObjectRef):
+
+    def __init__(self, opt_level=2, size_level=0):
```

Comment:
The `opt_level` and `size_level` arguments need to be validated - presently I can pass invalid values like `speed_level=4, size_level=-1` - with a release build this will result in junk values ending up somewhere being passed to the LLVM APIs, and I imagine the `llvm_unreachable`s in `mapLevel` above would end up killing the interpreter with a debug build. I think we need to check for validity using the same rules as in the `OptimizationLevel` constructor in LLVM's C++ implementation (speed level 0-3, size level 0-2, optimization for size requires speed level 2).